### PR TITLE
notebook with more instruction and some changes in scripts

### PR DESCRIPTION
**Pull request by 1000ping**
start_websocket_servers have some unnecessary lines to provide python versions and is causing errors on windows 10 but since in subprocess.Popen we provide the name of exe that will execute certain file with certain arguments. In run_websocket_client logging.debug is replaced by print because logging is not showing losses and errors. Notebook has more instructions and possible solutions for possible errors and removed some extra code

### CODE REVIEW

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 1]
```diff
@@ -1,15 +1,21 @@
 {
  "cells": [
   {
-   "cell_type": "markdown",
+   "cell_type": "code",
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: Should be markdown not code: have a look at your notebook in jupyter to make sure it prints nicely, here is how it looks like currently: https://github.com/OpenMined/PySyft/blob/039798359758a5f02d841c5d2c15fe676318a0c4/examples/tutorials/advanced/websockets-example-MNIST/Federated%20learning%20with%20websockets%20and%20federated%20averaging.ipynb

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 2]
```diff
@@ -20,7 +26,13 @@
     "\n",
     "Each worker is represented by two parts, a local handle (websocket client worker) and the remote instance that holds the data and performs the computations. The remote part is called a websocket server worker.\n",
     "\n",
-    "So first, we need to create the remote workers. For this, you need to run in a terminal (not possible from the notebook):\n",
+    "So first, you need to ```cd``` to the folder where this notebook and other additional files for running server and client are \n",
+    "\n",
+    "for example\n",
+    "\n",
+    "```cd PySyft-dev\\examples\\tutorials\\advanced\\websockets-example-MNIST```\n",
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: Don't put windows path
or make it clear when this is for windows users (you can use citations with '>' . in markdown)

Example:
> **For Windows users**
> Use `cd my\path`

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 3]
```diff
@@ -43,19 +55,39 @@
...
+   "execution_count": 2,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The autoreload extension is already loaded. To reload it, use:\n",
+      "  %reload_ext autoreload\n"
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: Remove this print

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 4]
```diff
@@ -142,12 +174,109 @@
     "\n",
     "We instantiate two FederatedDataLoaders, one for the train and one for the test set of the MNIST dataset.\n",
     "\n",
-    "*If you run into BrokenPipe errors try to restart the notebook or try running [run_websocket_client.py](run_websocket_client.py) (Python script that contains the same code of this notebook)*"
+    "*If you run into BrokenPipe errors go to the parrent directory of the directory where your project is and delete data folder then restart notebook and try again if the error comes again delete that data folder again run the following command*\n",
+    "for example \n",
+    "directory for data\n",
+    "```PySyft-dev/examples/tutorials/advanced/```\n",
+    "directory for project notebook and scripts\n",
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: It doesn't print nicely

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py [hunk 5]
```diff
@@ -68,7 +67,7 @@ def train_on_batches(worker, batches, model_in, device, lr):
         if batch_idx % LOG_INTERVAL == 0:
             loss = loss.get()  # <-- NEW: get the loss back
             loss_local = True
-            logger.debug(
+            print(
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: Why not keeping logger.debug?

File: examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py [hunk 6]
```diff
@@ -1,13 +1,12 @@
 import subprocess
-import sys
 
-python = "python" + sys.version[0:3]
 
-call_alice = [python, "run_websocket_server.py", "--port", "8777", "--id", "alice"]
+call_alice = ['python', "run_websocket_server.py", "--port", "8777", "--id", "alice"]
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py]: Why changing this?

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py (hunk 5 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: logger was not printing anything on the notebook output, I don't know what was causing that maybe it was because we were getting functions from script instead of running script directly from terminal so I tried print and loss after LOG_INTERVAL start showing in notebook

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb (hunk 1 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: I might have accidently did that while on pycharm changing it asap

Comment (LaRiffle) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: Install the markdown module for pycharm :)
Put you should prefer jupyter notebook I'd say :)

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb (hunk 2 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: done made it more generic

File: examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py (hunk 6 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py]: It was giving error no such file exists ... so I tested what arguments in subprocess.Popen are doing by using these arguments ['notepad','README.md'] so it opened README.md so what that line what doing is opening that README.md file with notepad.exe so in windows there is only one python.exe file in each envirenment there is no python3.exe or python3.6.exe so in windows 10 if I activate some conda environment to execute this script it will use python.exe of that environment for that subprocess.Popen  ... I dont know if it will behave the same on linux I havn't tested it on that if you want I can make an if for when using windows and when using linux

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py (hunk 5 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: but I will try to  do few more test on logging

File: examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py (hunk 6 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py]: ok so I tried writing !python3 and !python2 and !python3.6 on google colab and its working there so it means these will work on Linux so I am adding an if statement for people using Linux

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py (hunk 5 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: logging working now had to add steam handlers

**Comment by 1000ping** [comment]
@LaRiffle done with the changes you can check and let me know if any other changes thanks for pointing out my mistakes it was my first time starting a pulling request

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py [hunk 7]
```diff
@@ -121,7 +120,7 @@ def train(model, device, federated_train_loader, lr, federate_after_n_batches):
     counter = 0
 
     while True:
-        logger.debug("Starting training round, batches [%s, %s]", counter, counter + nr_batches)
+        logger.debug("Starting training round, batches [{}, {}]".format(counter, counter + nr_batches))
```
Comment (midokura-silvia) [inline examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: why the use of .format()?

File: examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py [hunk 8]
```diff
@@ -1,13 +1,24 @@
 import subprocess
 import sys
+import os
 
 python = "python" + sys.version[0:3]
```
Comment (midokura-silvia) [inline examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py]: so in short we need:
```
if os.name == "nt":
    python = "python"
else:
    python = "python" + sys.version[0:3]
```

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 9]
```diff
@@ -142,7 +168,109 @@
     "\n",
     "We instantiate two FederatedDataLoaders, one for the train and one for the test set of the MNIST dataset.\n",
     "\n",
-    "*If you run into BrokenPipe errors try to restart the notebook or try running [run_websocket_client.py](run_websocket_client.py) (Python script that contains the same code of this notebook)*"
```
Comment (midokura-silvia) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: I think we can remove the comment concerning the BrokenPipe issue. Should be resolved by now (improved startup parameters of the websocket connection).

File: examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py (hunk 7 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py]: @midokura-silvia no particular reason to me it was more pythonic approach also these in .format approach you don't need to specify any type I can change it back if you think there is no need for change

File: examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py (hunk 8 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/start_websocket_servers.py]: true will make that change

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb (hunk 9 above)
Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: this error happened to me 2 days ago when I was trying to run that box. When was the issue resolved?

Comment (midokura-silvia) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: I don't really use the notebooks, so it seems that I confused it with another problem. If it persists, leave it in.

File: examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb [hunk 10]
```diff
@@ -209,63 +337,20 @@
     "print(model)"
    ]
   },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "Now we are ready to define the training loop. We will perform training over a given number of batches separately on each worker and then calculate the federated average of the resulting model."
```
Comment (LaRiffle) [inline examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: Maybe keep this information even if you remove the train function description, because it is important to understand how the training works.

Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: sure

Comment (1000ping) [reply examples/tutorials/advanced/websockets-example-MNIST/Federated learning with websockets and federated averaging.ipynb]: @LaRiffle done
